lib/chibi/math/prime.scm

```diff
@@ -175,57 +189,105 @@
+;; \var{put-1} for \var{p}=-1, called as \scheme{(\var{put-1} \var{r})}).
+(define (make-factorizer r1 put . o)
+  (let-optionals o ((put2  (lambda (r k) (put r  2 k)))
+                    (put-1 (lambda (r)   (put r -1 1))))
+    (lambda (n)
+      (when (zero? n)
+        (error "cannot factor 0"))
+      (factor-twos
+       n
+       (lambda (k2 n)
+         (let lp ((i 3) (ii 9)
+                  (n (abs n))
+                  (res (let ((res (if (negative? n) (put-1 r1) r1)))
+                         (if (zero? k2) res (put2 res k2)))))
+           (let next-i ((i i) (ii ii))
+             (cond ((> ii n)
+                    (if (= n 1) res (put res n 1)))
+                   ((not (zero? (remainder n i)))
+                    (next-i (+ i 2) (+ ii (* (+ i 1) 4))))
+                   (else
+                    (let rest ((n (quotient n i))
+                               (k 1))
+                      (if (zero? (remainder n i))
+                          (rest (quotient n i) (+ k 1))
+                          (lp (+ i 2) (+ ii (* (+ i 1) 4))
+                              n (put res i k)))))))))))))
+
+;;> Returns the factorization of \var{n} as a list of
+;;> elements of the form \scheme{(\var{p} . \var{k})},
+;;> where \var{p} is a prime factor
+;;> and \var{k} is its multiplicity.
+(define factor-alist
+  (let ((rfactor (make-factorizer '()
+                  (lambda (l p k) (cons (cons p k) l)))))
+    (lambda (n) (reverse (rfactor n)))))
+
 ;;> Returns the factorization of \var{n} as a monotonically
 ;;> increasing list of primes.
 (define (factor n)
-  (cond
-   ((negative? n)
-    (cons -1 (factor (- n))))
-   ((<= n 2)
-    (list n))
-   (else
-    (let lp ((n n)
-             (res (list)))
-      (cond
-       ((even? n)
-        (lp (quotient n 2) (cons 2 res)))
-       ((= n 1)
-        (reverse res))
-       (else
-        (let lp ((i 3) (n n) (limit (exact (ceiling (sqrt n)))) (res res))
-          (cond
-           ((= n 1)
-            (reverse res))
-           ((> i limit)
-            (reverse (cons n res)))
-           ((zero? (remainder n i))
-            (lp i (quotient n i) limit (cons i res)))
-           (else
-            (lp (+ i 2) n limit res))))))))))
+  (when (zero? n)
+    (error "cannot factor 0"))
+  (factor-twos
+   n
+   (lambda (b n)
+     (let lp ((i 3) (ii 9) (n (abs n))
+              (res (append (make-list b 2)
+                           (if (negative? n) (list -1) '()))))
+       (cond
+        ((> ii n)
+         (reverse (if (= n 1) res (cons n res))))
+        ((zero? (remainder n i))
+         (lp i ii (quotient n i) (cons i res)))
+        (else
+         (lp (+ i 2)
+             (+ ii (* (+ i 1) 4))
+             n
+             res)))))))
+;; this version is slightly slower
+;;(define factor
+;;  (let ((rfactor (make-factorizer '()
+;;                  (lambda (l p k) (cons (make-list k p) l)))))
+;;    (lambda (n) (apply append! (reverse (rfactor n))))))
```

Comment:
I think I should either 

1. kill these lines (if speed is all that matters) or 
2. swap this version in for the one above (if size matters more and a 5% slowdown is no big deal)

Reply:
At a 5% diff I'd prefer the smaller size here for Chibi.

Reply:
ok,  done.